internal/authz/authz.go

```diff
@@ -0,0 +1,434 @@
+	if err != nil {
+		return nil, err
+	}
+
+	storeIDs := make([]string, len(resp.GetObjects()))
+	storePrefix := fmt.Sprintf("%s:", StoreType)
+	for i, store := range resp.GetObjects() {
+		storeIDs[i] = strings.TrimPrefix(store, storePrefix)
+	}
+
+	return storeIDs, nil
+}
+
+// GetModulesForWriteRequest returns the modules that should be checked for the write request.
+// If we encounter a type with no attached module, we should break and return no modules so that the authz check will be against the store
+// Otherwise we return a list of unique modules encountered so that FGA on FGA can check them after.
+func (a *Authorizer) GetModulesForWriteRequest(req *openfgav1.WriteRequest, typesys *typesystem.TypeSystem) ([]string, error) {
+	tuples := make([]TupleKeyInterface, len(req.GetWrites().GetTupleKeys())+len(req.GetDeletes().GetTupleKeys()))
+	var index int
+	for _, tuple := range req.GetWrites().GetTupleKeys() {
+		tuples[index] = tuple
+		index++
+	}
+	for _, tuple := range req.GetDeletes().GetTupleKeys() {
+		tuples[index] = tuple
+		index++
+	}
+
+	modulesMap, err := extractModulesFromTuples(tuples, typesys)
+	if err != nil {
+		return nil, err
+	}
+
+	modules := make([]string, len(modulesMap))
+	var i int
+	for module := range modulesMap {
+		modules[i] = module
+		i++
+	}
+
+	return modules, nil
+}
+
+// TupleKeyInterface is an interface that both TupleKeyWithoutCondition and TupleKey implement.
+type TupleKeyInterface interface {
+	GetObject() string
+	GetRelation() string
+}
+
+// extractModulesFromTuples extracts the modules from the tuples. If a type has no module, we
+// return an empty map so that the caller can handle authorization for tuples without modules.
+func extractModulesFromTuples[T TupleKeyInterface](tupleKeys []T, typesys *typesystem.TypeSystem) (map[string]struct{}, error) {
+	modulesMap := make(map[string]struct{})
+	for _, tupleKey := range tupleKeys {
+		objType, _ := tuple.SplitObject(tupleKey.GetObject())
+		objectType, ok := typesys.GetTypeDefinition(objType)
+		if !ok {
+			return nil, serverErrors.TypeNotFound(objType)
+		}
+		module, err := parser.GetModuleForObjectTypeRelation(objectType, tupleKey.GetRelation())
+		if err != nil {
+			return nil, err
+		}
+		if module == "" {
+			return nil, nil
+		}
+		modulesMap[module] = struct{}{}
+	}
+
+	return modulesMap, nil
+}
+
+func (a *Authorizer) individualAuthorize(ctx context.Context, clientID, relation, object string, contextualTuples *openfgav1.ContextualTupleKeys) error {
+	req := &openfgav1.CheckRequest{
+		StoreId:              a.config.StoreID,
+		AuthorizationModelId: a.config.ModelID,
+		TupleKey: &openfgav1.CheckRequestTupleKey{
+			User:     ClientIDType(clientID).String(),
+			Relation: relation,
+			Object:   object,
+		},
+		ContextualTuples: contextualTuples,
+	}
+
+	// Disable authz check for the check request.
+	ctx = authclaims.ContextWithSkipAuthzCheck(ctx, true)
+	resp, err := a.server.Check(ctx, req)
+	if err != nil {
+		return err
```

Comment:
(not blocking)

```suggestion
		return fmt.Errorf("error authorizing the store: %w", err)
```

this makes it easier for me to identify the source of the problem. Compare 

```
$ fga store --config root.config.yaml --store-id="$PROJECT_STORE_ID" get

Error: failed to get store 01J9YWE5HEDR3X0AABYFA8YAHD due to GetStore validation error for GET GetStore with body {"code":"authorization_model_not_found","message":"Authorization Model '01J9VG313KCD6573NZM80ANPET' not found"}
 with error code authorization_model_not_found error message: Authorization Model '01J9VG313KCD6573NZM80ANPET' not found
```

this gives me no indication that the error occurred during the authz call. Versus

```
Error: failed to get store 01J9YWE5HEDR3X0AABYFA8YAHD due to GetStore validation error for GET GetStore with body {"code":"authorization_model_not_found","message":"error authorizing the store: Authorization Model '01J9YWE8DR2HS2EQP5JFB05W0C' not found"}
 with error code authorization_model_not_found error message: error authorizing the store: Authorization Model '01J9YWE8DR2HS2EQP5JFB05W0C' not found
```

Reply:
Putting this here is resulting in nested `error authorizing the call: error authorizing the call:...` error messages. I will refactor a bit so that we can have this context without the nesting